src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckWorker.scala

```diff
@@ -3,124 +3,115 @@ package core.health.impl
 
 import java.net.{InetSocketAddress, Socket}
 import java.security.cert.X509Certificate
-import javax.net.ssl.{KeyManager, SSLContext, X509TrustManager}
 
-import akka.actor.{Actor, ActorSystem, PoisonPill}
-import akka.http.scaladsl.settings.ClientConnectionSettings
+import akka.actor.ActorSystem
 import akka.http.scaladsl.client.RequestBuilding
 import akka.http.scaladsl.model.{HttpRequest, HttpResponse, headers}
+import akka.http.scaladsl.settings.ClientConnectionSettings
 import akka.http.scaladsl.{ConnectionContext, Http}
-import akka.stream.Materializer
+import akka.stream.ActorMaterializer
 import akka.stream.scaladsl.{Sink, Source}
 import com.typesafe.scalalogging.StrictLogging
 import com.typesafe.sslconfig.akka.AkkaSSLConfig
+import javax.net.ssl.{KeyManager, SSLContext, X509TrustManager}
+import mesosphere.marathon.Protos.HealthCheckDefinition.Protocol
 import mesosphere.marathon.core.health._
 import mesosphere.marathon.core.instance.Instance
 import mesosphere.marathon.state.{AppDefinition, Timestamp}
 import mesosphere.util.ThreadPoolContext
+import mesosphere.marathon.util.toRichFuture
 
 import scala.concurrent.Future
 import scala.concurrent.duration.FiniteDuration
 import scala.util.control.NonFatal
 import scala.util.{Failure, Success}
 
-class HealthCheckWorkerActor(implicit mat: Materializer) extends Actor with StrictLogging {
-
-  import HealthCheckWorker._
+object HealthCheckWorker extends StrictLogging {
 
-  private implicit val system = context.system
-  import context.dispatcher // execution context for futures
+  def run(app: AppDefinition, instance: Instance, healthCheck: MarathonHealthCheck)(implicit mat: ActorMaterializer): Future[HealthResult] = {
+    logger.debug("Dispatching health check job for {}", instance.instanceId)
 
-  def receive: Receive = {
-    case HealthCheckJob(app, instance, check) =>
-      val replyTo = sender() // avoids closing over the volatile sender ref
+    implicit val system = mat.system
+    implicit val ex = mat.executionContext
 
-      doCheck(app, instance, check)
-        .andThen {
-          case Success(Some(result)) => replyTo ! result
-          case Success(None) => // ignore
-          case Failure(t) =>
-            logger.warn(s"Performing health check for app=${app.id} instance=${instance.instanceId} port=${check.port} failed with exception", t)
-            replyTo ! Unhealthy(
-              instance.instanceId,
-              instance.runSpecVersion,
-              s"${t.getClass.getSimpleName}: ${t.getMessage}"
-            )
-        }
-        .onComplete { _ => self ! PoisonPill }
+    check(app, instance, healthCheck)
+      .mapAll {
```

Comment:
This is Scala 2.12 so you can replace this custom `mapAll` with proper `transform` and also remove it below in `RichFuture`